velox/exec/tests/utils/QueryAssertions.cpp

```diff
@@ -703,9 +775,135 @@ std::shared_ptr<Task> assertQueryReturnsEmptyResult(
   return result.first->task();
 }
 
-// Special function to compare multisets with vectors of variants in a way that
-// we compare all floating point values inside using 'epsilon' constant.
-// Returns true if equal.
+// Tranform a MaterializedRow vector by moving floating-point columns to the end
+// of every row. The number of columns per row and the position of
+// floating-point columns are given by width and floatColumnIndices.
+void moveFloatColumnsToEnd(
+    const std::vector<velox::column_index_t>& floatColumnIndices,
+    size_t width,
+    std::vector<MaterializedRow>& data) {
+  for (auto i = 0; i < data.size(); ++i) {
+    velox::column_index_t lastNonFloatColumn = width - 1;
+    for (auto j = (int32_t)floatColumnIndices.size() - 1; j >= 0; --j) {
+      if (floatColumnIndices[j] != lastNonFloatColumn) {
+        std::swap(data[i][floatColumnIndices[j]], data[i][lastNonFloatColumn]);
+      }
+      --lastNonFloatColumn;
+    }
+  }
+}
+
+// Return true if currentRow and previousRow are the same at the first width
+// columns.
+bool equalPrefix(
+    const MaterializedRow& currentRow,
+    const MaterializedRow& previousRow,
+    size_t width) {
+  for (auto i = 0; i < width; ++i) {
+    if (currentRow[i] != previousRow[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+
+// Return true if there are multiple rows with the same prefix of the first
+// width columns. The argument groupedRows should be sorted by the first
+// width columns.
+bool hasMultiRowGroups(
+    const std::vector<MaterializedRow>& sortedRows,
+    size_t width) {
+  for (auto i = 1; i < sortedRows.size(); ++i) {
+    if (equalPrefix(sortedRows[i], sortedRows[i - 1], width)) {
+      return true;
+    }
+  }
+  return false;
+}
+
+std::vector<velox::column_index_t> getFloatColumnIndices(const MaterializedRowMultiset& left,
+    const MaterializedRowMultiset& right) {
+  std::vector<velox::column_index_t> floatColumnIndices;
+  auto& leftRow = *left.begin();
+  auto& rightRow = *right.begin();
+  auto width = leftRow.size();
+  for (auto i = 0; i < width; ++i) {
+    if (leftRow[i].kind() != rightRow[i].kind()) {
+      VELOX_FAIL("Expect the same type between two result sets.");
+    }
+    if ((leftRow[i].kind() == TypeKind::REAL ||
+         leftRow[i].kind() == TypeKind::DOUBLE)) {
+      floatColumnIndices.push_back(i);
+    }
+  }
+  return floatColumnIndices;
+}
+
+// Sort rows in left and right into leftSorted and rightSorted
+// and returns whether the sorting by unique keys is successful. The sorting is
+// successful if the following requirements are satisfied.
+//   1. left and right are non-empty and have the same number of columns of the
```

Comment:
I feel that this check doesn't belong to this function. This function should just assume that this is the case, no?